### DISCUSSION TITLE
delete-merged

### DISCUSSION THREAD

**Discussion by raylu** [VirtusLab/git-machete#1088] [Ideas]
(continuing from https://github.com/VirtusLab/git-machete/issues/1057#issuecomment-1797030780)

I saw `delete-unmanaged` when I first read the help, but deleting all unmanaged branches seems scary to me. I only use machete when stacking a PR. for regular branches off of `main`, I generally don't bother, so my machete status is often out of date

for `github sync`, I really can't imagine when I would want to add local branches if I don't already have them. I do totally have multiple clones of repos (in different environments), but if I didn't already make a local copy of a branch, I generally don't want it

would you be open adding a `delete-merged` command that only deletes branches if
1. they have a remote tracking branch (don't delete unpushed branches)
2. the remote tracking branch doesn't exist (only delete merged branches)
3. there are no children/stacks built on top of them

I considered adding a flag like `delete-unamanged --remote-deleted-only`, but this _would_ delete branches even if they are managed

I'm happy to write the PR if you want

**Comment by PawelLipski** [comment]
Hmmm just from a quick look:

> they have a remote tracking branch (don't delete unpushed branches)
> the remote tracking branch doesn't exist (only delete merged branches)

Is this apparent contradiction explained b/c you mean _remote tracking branch_ once as _an entry in .git/refs/remotes locally_  and once _a branch (aka an entry in .git/refs/heads) in remote repo_? 🤔 

> I considered adding a flag like delete-unamanged --remote-deleted-only, but this would delete branches even if they are managed

I've thought of `slide-out --untracked` (https://github.com/VirtusLab/git-machete/issues/1051) 🤔 

> I'm happy to write the PR if you want

Wow I'd appreciate that a lot... I slowed down contributions to OSS recently as I've got a lot of work in commercial projects. Let's first smooth out the details ofc

**Comment by raylu** [comment]
>> they have a remote tracking branch (don't delete unpushed branches)
>> the remote tracking branch doesn't exist (only delete merged branches)

> Is this apparent contradiction explained b/c you mean _remote tracking branch_ once as _an entry in .git/refs/remotes locally_ and once _a branch (aka an entry in .git/refs/heads) in remote repo_? 🤔

I mean they have `branch.<name>.remote` and `branch.<name>.merge` configuration entries (as in `git branch --track`) but the branch they are tracking doesn't exist (`git branch --remotes` doesn't show it because `.git/logs/refs/remotes/origin/<name>` doesn't exist)

basically, the branch has been deleted on github (by the "automatically delete head branches" setting), your local clone acknowledges it's gone (after `git fetch --prune`), but you still have a useless local branch

**Reply by PawelLipski** [reply]
Okay! makes sense indeed

**Reply by PawelLipski** [reply]
This would be way better than what `git machete clean` does right now (just remove untracked branches, no matter whether its tracking config exists)

**Reply by raylu** [reply]
here's what I've got so far: https://github.com/raylu/git-machete/compare/develop...raylu:git-machete:clean. just need to make the tests delete remote branches (or set the config to have a remote branch without pushing)

**Reply by PawelLipski** [reply]
Wow nice! Pls open a draft PR for that... I'd rather leave `clean` unchanged & deprecated, TBD where to put this new functionality of yours

**Reply by raylu** [reply]
#1094

**Comment by PawelLipski** [comment]
Done, available as `git machete slide-out --removed-from-remote`
